### ISSUE TITLE
Stack is unable to use ghc 8.0.1 or ghc 8.4.4

### ISSUE THREAD

**Issue by rsolano1-uw** [commercialhaskell/stack#5240]
### General summary/comments
I set stack to use system-ghc, but it does not use system-ghc

### Steps to reproduce

1. Have a system-installed version of GHC
2. Download stack binary for manual installation
3. Run `stack setup` and watch it error out because `No setup information found for ghc-8.4.3 on your platform.`
4. Edit the file `~/.stack/config.yaml` to have the line `system-ghc: true` at the end
5. `stack path`

### Expected

Stack prints some output about my system-installed version of GHC

### Actual

With my system-installed version of ghc-8.0.1
```
$ stack path --verbose
Version 2.1.3, Git revision 636e3a759d51127df2b62f90772def126cdf6d1f (7735 commits) x86_64 hpack-0.31.2
2020-03-24 14:39:31.040720: [debug] Checking for project config at: /net/gs/vol1/home/rsolano1/stack.yaml
2020-03-24 14:39:31.042043: [debug] Checking for project config at: /net/gs/vol1/home/stack.yaml
2020-03-24 14:39:31.042163: [debug] Checking for project config at: /net/gs/vol1/stack.yaml
2020-03-24 14:39:31.042365: [debug] Checking for project config at: /net/gs/stack.yaml
2020-03-24 14:39:31.177515: [debug] Checking for project config at: /net/stack.yaml
2020-03-24 14:39:31.182524: [debug] Checking for project config at: /stack.yaml
2020-03-24 14:39:31.182626: [debug] No project config file found, using defaults.
2020-03-24 14:39:31.214719: [debug] SELECT COUNT(*) FROM "last_performed" WHERE ("action"=?) AND ("timestamp">=?); [PersistInt64 1,PersistUTCTime 2020-03-23 21:39:31.213545749 UTC]
2020-03-24 14:39:31.242065: [debug] Run from outside a project, using implicit global project config
2020-03-24 14:39:31.244336: [debug] Using resolver: lts-15.5 from implicit global project's config file: /net/gs/vol1/home/rsolano1/.stack/global-project/stack.yaml
2020-03-24 14:39:31.244792: [debug] Using package location completions from a lock file
2020-03-24 14:39:31.687142: [warn] Stack has not been tested with GHC versions above 8.6, and using 8.8.3, this may fail
2020-03-24 14:39:31.687297: [debug] Getting system compiler version
2020-03-24 14:39:31.695889: [debug] SELECT "id","actual_version","arch","ghc_path","ghc_size","ghc_modified","ghc_pkg_path","runghc_path","haddock_path","cabal_version","global_db","global_db_cache_size","global_db_cache_modified","info","global_dump" FROM "compiler_cache" WHERE "ghc_path"=?; [PersistText "/net/gs/vol3/software/modules-sw/ghc/8.0.1/Linux/RHEL6/x86_64/bin/ghc-8.0.1"]
2020-03-24 14:39:31.809878: [debug] Loaded compiler information from cache
2020-03-24 14:39:31.810316: [debug] Not using compiler at "/net/gs/vol3/software/modules-sw/ghc/8.0.1/Linux/RHEL6/x86_64/bin/ghc-8.0.1": Control.Exception.Safe.throwString called with:

Not the compiler version we want
Called from:
  throwString (src/Stack/Setup.hs:625:55 in stack-2.1.3-HEoMsL9gvgt2vNexTWWwGx:Stack.Setup)

2020-03-24 14:39:31.812784: [debug] SELECT "id","actual_version","arch","ghc_path","ghc_size","ghc_modified","ghc_pkg_path","runghc_path","haddock_path","cabal_version","global_db","global_db_cache_size","global_db_cache_modified","info","global_dump" FROM "compiler_cache" WHERE "ghc_path"=?; [PersistText "/usr/bin/ghc-7.0.4"]
2020-03-24 14:39:31.816175: [debug] Looking for executable(s): ["/usr/bin/ghc-pkg-7.0.4","/usr/bin/ghc-pkg"]
2020-03-24 14:39:31.816320: [debug] Looking for executable(s): ["/usr/bin/runghc-7.0.4","/usr/bin/runghc"]
2020-03-24 14:39:31.816419: [debug] Looking for executable(s): ["/usr/bin/haddock-7.0.4","/usr/bin/haddock-ghc-7.0.4","/usr/bin/haddock","/usr/bin/haddock-ghc"]
2020-03-24 14:39:31.816599: [debug] Run process: /usr/bin/ghc-7.0.4 --info
2020-03-24 14:39:31.840480: [debug] Process finished in 24ms: /usr/bin/ghc-7.0.4 --info
2020-03-24 14:39:31.841812: [debug] Run process: /usr/bin/ghc-pkg-7.0.4 --global --no-user-package-db dump --expand-pkgroot
2020-03-24 14:39:31.902583: [debug] Not using compiler at "/usr/bin/ghc-7.0.4": Found an invalid compiler at "/usr/bin/ghc-7.0.4": Received ExitFailure 1 when running
Raw command: /usr/bin/ghc-pkg-7.0.4 --global --no-user-package-db dump --expand-pkgroot
Standard error:

ghc-pkg: unrecognized option `--no-user-package-db'
unrecognized option `--expand-pkgroot'
Usage:
  ghc-pkg init {path}
    Create and initialise a package database at the location {path}.
    Packages can be registered in the new database using the register
    command with --package-conf={path}.  To use the new database with GHC,
    use GHC's -package-conf flag.

  ghc-pkg register {filename | -}
    Register the package using the specified installed package
    description. The syntax for the latter is given in the ghc-pkg
    documentation.  The input file should be encoded in UTF-8.

  ghc-pkg update {filename | -}
    Register the package, overwriting any other package with the
    same name. The input file should be encoded in UTF-8.

  ghc-pkg unregister {pkg-id}
    Unregister the specified package.

  ghc-pkg expose {pkg-id}
    Expose the specified package.

  ghc-pkg hide {pkg-id}
    Hide the specified package.

  ghc-pkg list [pkg]
    List registered packages in the global database, and also the
    user database if --user is given. If a package name is given
    all the registered versions will be listed in ascending order.
    Accepts the --simple-output flag.

  ghc-pkg dot
    Generate a graph of the package dependencies in a form suitable
    for input for the graphviz tools.  For example, to generate a PDF    of the dependency graph: ghc-pkg dot | tred | dot -Tpdf >pkgs.pdf
  ghc-pkg find-module {module}
    List registered packages exposing module {module} in the global
    database, and also the user database if --user is given.
    All the registered versions will be listed in ascending order.
    Accepts the --simple-output flag.

  ghc-pkg latest {pkg-id}
    Prints the highest registered version of a package.

  ghc-pkg check
    Check the consistency of package depenencies and list broken packages.
    Accepts the --simple-output flag.

  ghc-pkg describe {pkg}
    Give the registered description for the specified package. The
    description is returned in precisely the syntax required by ghc-pkg
    register.

  ghc-pkg field {pkg} {field}
    Extract the specified field of the package description for the
    specified package. Accepts comma-separated multiple fields.

  ghc-pkg dump
    Dump the registered description for every package.  This is like
    "ghc-pkg describe '*'", except that it is intended to be used
    by tools that parse the results, rather than humans.  The output is
    always encoded in UTF-8, regardless of the current locale.

  ghc-pkg recache
    Regenerate the package database cache.  This command should only be
    necessary if you added a package to the database by dropping a file
    into the database directory manually.  By default, the global DB
    is recached; to recache a different DB use --user or --package-conf
    as appropriate.

 Substring matching is supported for {module} in find-module and
 for {pkg} in list, describe, and field, where a '*' indicates
 open substring ends (prefix*, *suffix, *infix*).

  When asked to modify a database (register, unregister, update,
  hide, expose, and also check), ghc-pkg modifies the global database by
  default.  Specifying --user causes it to act on the user database,
  or --package-conf can be used to act on another database
  entirely. When multiple of these options are given, the rightmost
  one is used as the database to act upon.

  Commands that query the package database (list, tree, latest, describe,
  field) operate on the list of databases specified by the flags
  --user, --global, and --package-conf.  If none of these flags are
  given, the default is --global --user.

 The following optional flags are also accepted:

                 --user                  use the current user's package database
                 --global                use the global package database
  -f FILE        --package-conf=FILE     use the specified package config file
                 --global-conf=FILE      location of the global package config
                 --no-user-package-conf  never read the user package database
                 --force                 ignore missing dependencies, directories, and libraries
                 --force-files           ignore missing directories and libraries only
  -g             --auto-ghci-libs        automatically build libs for GHCi (with register)
  -?             --help                  display this help and exit
  -V             --version               output version information and exit
                 --simple-output         print output in easy-to-parse format for some commands
                 --names-only            only print package names, not versions; can only be used with list --simple-output
                 --ignore-case           ignore case for substring matching
  -v[Verbosity]  --verbose[=Verbosity]   verbosity level (0-2, default 1)


2020-03-24 14:39:31.904625: [debug] Installed tools: 
 - 
2020-03-24 14:39:31.904796: [debug] Run process: /sbin/ldconfig -p
2020-03-24 14:39:31.909099: [debug] Process finished in 4ms: /sbin/ldconfig -p
2020-03-24 14:39:31.909494: [debug] Found shared library libtinfo.so.5 in 'ldconfig -p' output
2020-03-24 14:39:31.910853: [debug] Did not find shared library libtinfo.so.6
2020-03-24 14:39:31.911213: [debug] Did not find shared library libncursesw.so.6
2020-03-24 14:39:31.911491: [debug] Did not find shared library libgmp.so.10
2020-03-24 14:39:31.911643: [debug] Found shared library libgmp.so.3 in 'ldconfig -p' output
2020-03-24 14:39:31.911795: [debug] Potential GHC builds: gmp4
2020-03-24 14:39:31.911876: [debug] Found already installed GHC builds: 
2020-03-24 14:39:32.234959: [debug] Trying to setup GHC build: gmp4
2020-03-24 14:39:32.235985: [error] No setup information found for ghc-8.8.3 on your platform.
This probably means a GHC bindist has not yet been added for OS key 'linux64-gmp4'.
Supported versions: ghc-7.8.4, ghc-7.10.2, ghc-7.10.3, ghc-8.0.1, ghc-8.0.2, ghc-8.2.1, ghc-8.2.2, ghc-8.4.2
```

I get different errors in --verbose if I change my path to include the ghc-8.4.4 installation instead of the ghc-8.0.1 installation. That comes out to:

```
$ stack path --verbose
Version 2.1.3, Git revision 636e3a759d51127df2b62f90772def126cdf6d1f (7735 commits) x86_64 hpack-0.31.2
2020-03-24 14:44:28.800553: [debug] Checking for project config at: /net/gs/vol1/home/rsolano1/stack.yaml
2020-03-24 14:44:28.800831: [debug] Checking for project config at: /net/gs/vol1/home/stack.yaml
2020-03-24 14:44:28.800910: [debug] Checking for project config at: /net/gs/vol1/stack.yaml
2020-03-24 14:44:28.800983: [debug] Checking for project config at: /net/gs/stack.yaml
2020-03-24 14:44:28.938021: [debug] Checking for project config at: /net/stack.yaml
2020-03-24 14:44:28.943117: [debug] Checking for project config at: /stack.yaml
2020-03-24 14:44:28.943222: [debug] No project config file found, using defaults.
2020-03-24 14:44:29.001569: [debug] SELECT COUNT(*) FROM "last_performed" WHERE ("action"=?) AND ("timestamp">=?); [PersistInt64 1,PersistUTCTime 2020-03-23 21:44:29.000600602 UTC]
2020-03-24 14:44:29.010139: [debug] Run from outside a project, using implicit global project config
2020-03-24 14:44:29.012533: [debug] Using resolver: lts-15.5 from implicit global project's config file: /net/gs/vol1/home/rsolano1/.stack/global-project/stack.yaml
2020-03-24 14:44:29.012996: [debug] Using package location completions from a lock file
2020-03-24 14:44:29.485466: [warn] Stack has not been tested with GHC versions above 8.6, and using 8.8.3, this may fail
2020-03-24 14:44:29.485621: [debug] Getting system compiler version
2020-03-24 14:44:29.533680: [debug] SELECT "id","actual_version","arch","ghc_path","ghc_size","ghc_modified","ghc_pkg_path","runghc_path","haddock_path","cabal_version","global_db","global_db_cache_size","global_db_cache_modified","info","global_dump" FROM "compiler_cache" WHERE "ghc_path"=?; [PersistText "/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/ghc-8.4.4"]
2020-03-24 14:44:29.537430: [debug] Looking for executable(s): ["/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/ghc-pkg-8.4.4","/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/ghc-pkg"]
2020-03-24 14:44:29.538256: [debug] Looking for executable(s): ["/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/runghc-8.4.4","/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/runghc"]
2020-03-24 14:44:29.538684: [debug] Looking for executable(s): ["/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/haddock-8.4.4","/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/haddock-ghc-8.4.4","/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/haddock","/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/haddock-ghc"]
2020-03-24 14:44:29.540310: [debug] Not using compiler at "/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/ghc-8.4.4": Found an invalid compiler at "/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/ghc-8.4.4": Control.Exception.Safe.throwString called with:

Could not find any of: ["/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/haddock-8.4.4","/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/haddock-ghc-8.4.4","/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/haddock","/net/gs/vol3/software/modules-sw/ghc/8.4.4/Linux/RHEL6/x86_64/bin/haddock-ghc"]
Called from:
  throwString (src/Stack/Setup.hs:721:25 in stack-2.1.3-HEoMsL9gvgt2vNexTWWwGx:Stack.Setup)

2020-03-24 14:44:29.542938: [debug] SELECT "id","actual_version","arch","ghc_path","ghc_size","ghc_modified","ghc_pkg_path","runghc_path","haddock_path","cabal_version","global_db","global_db_cache_size","global_db_cache_modified","info","global_dump" FROM "compiler_cache" WHERE "ghc_path"=?; [PersistText "/usr/bin/ghc-7.0.4"]
2020-03-24 14:44:29.546126: [debug] Looking for executable(s): ["/usr/bin/ghc-pkg-7.0.4","/usr/bin/ghc-pkg"]
2020-03-24 14:44:29.546267: [debug] Looking for executable(s): ["/usr/bin/runghc-7.0.4","/usr/bin/runghc"]
2020-03-24 14:44:29.546363: [debug] Looking for executable(s): ["/usr/bin/haddock-7.0.4","/usr/bin/haddock-ghc-7.0.4","/usr/bin/haddock","/usr/bin/haddock-ghc"]
2020-03-24 14:44:29.546541: [debug] Run process: /usr/bin/ghc-7.0.4 --info
2020-03-24 14:44:29.576154: [debug] Process finished in 30ms: /usr/bin/ghc-7.0.4 --info
2020-03-24 14:44:29.577551: [debug] Run process: /usr/bin/ghc-pkg-7.0.4 --global --no-user-package-db dump --expand-pkgroot
2020-03-24 14:44:29.639290: [debug] Not using compiler at "/usr/bin/ghc-7.0.4": Found an invalid compiler at "/usr/bin/ghc-7.0.4": Received ExitFailure 1 when running
Raw command: /usr/bin/ghc-pkg-7.0.4 --global --no-user-package-db dump --expand-pkgroot
Standard error:

ghc-pkg: unrecognized option `--no-user-package-db'
unrecognized option `--expand-pkgroot'
Usage:
  ghc-pkg init {path}
    Create and initialise a package database at the location {path}.
    Packages can be registered in the new database using the register
    command with --package-conf={path}.  To use the new database with GHC,
    use GHC's -package-conf flag.

  ghc-pkg register {filename | -}
    Register the package using the specified installed package
    description. The syntax for the latter is given in the ghc-pkg
    documentation.  The input file should be encoded in UTF-8.

  ghc-pkg update {filename | -}
    Register the package, overwriting any other package with the
    same name. The input file should be encoded in UTF-8.

  ghc-pkg unregister {pkg-id}
    Unregister the specified package.

  ghc-pkg expose {pkg-id}
    Expose the specified package.

  ghc-pkg hide {pkg-id}
    Hide the specified package.

  ghc-pkg list [pkg]
    List registered packages in the global database, and also the
    user database if --user is given. If a package name is given
    all the registered versions will be listed in ascending order.
    Accepts the --simple-output flag.

  ghc-pkg dot
    Generate a graph of the package dependencies in a form suitable
    for input for the graphviz tools.  For example, to generate a PDF    of the dependency graph: ghc-pkg dot | tred | dot -Tpdf >pkgs.pdf
  ghc-pkg find-module {module}
    List registered packages exposing module {module} in the global
    database, and also the user database if --user is given.
    All the registered versions will be listed in ascending order.
    Accepts the --simple-output flag.

  ghc-pkg latest {pkg-id}
    Prints the highest registered version of a package.

  ghc-pkg check
    Check the consistency of package depenencies and list broken packages.
    Accepts the --simple-output flag.

  ghc-pkg describe {pkg}
    Give the registered description for the specified package. The
    description is returned in precisely the syntax required by ghc-pkg
    register.

  ghc-pkg field {pkg} {field}
    Extract the specified field of the package description for the
    specified package. Accepts comma-separated multiple fields.

  ghc-pkg dump
    Dump the registered description for every package.  This is like
    "ghc-pkg describe '*'", except that it is intended to be used
    by tools that parse the results, rather than humans.  The output is
    always encoded in UTF-8, regardless of the current locale.

  ghc-pkg recache
    Regenerate the package database cache.  This command should only be
    necessary if you added a package to the database by dropping a file
    into the database directory manually.  By default, the global DB
    is recached; to recache a different DB use --user or --package-conf
    as appropriate.

 Substring matching is supported for {module} in find-module and
 for {pkg} in list, describe, and field, where a '*' indicates
 open substring ends (prefix*, *suffix, *infix*).

  When asked to modify a database (register, unregister, update,
  hide, expose, and also check), ghc-pkg modifies the global database by
  default.  Specifying --user causes it to act on the user database,
  or --package-conf can be used to act on another database
  entirely. When multiple of these options are given, the rightmost
  one is used as the database to act upon.

  Commands that query the package database (list, tree, latest, describe,
  field) operate on the list of databases specified by the flags
  --user, --global, and --package-conf.  If none of these flags are
  given, the default is --global --user.

 The following optional flags are also accepted:

                 --user                  use the current user's package database
                 --global                use the global package database
  -f FILE        --package-conf=FILE     use the specified package config file
                 --global-conf=FILE      location of the global package config
                 --no-user-package-conf  never read the user package database
                 --force                 ignore missing dependencies, directories, and libraries
                 --force-files           ignore missing directories and libraries only
  -g             --auto-ghci-libs        automatically build libs for GHCi (with register)
  -?             --help                  display this help and exit
  -V             --version               output version information and exit
                 --simple-output         print output in easy-to-parse format for some commands
                 --names-only            only print package names, not versions; can only be used with list --simple-output
                 --ignore-case           ignore case for substring matching
  -v[Verbosity]  --verbose[=Verbosity]   verbosity level (0-2, default 1)


2020-03-24 14:44:29.641282: [debug] Installed tools: 
 - 
2020-03-24 14:44:29.641472: [debug] Run process: /sbin/ldconfig -p
2020-03-24 14:44:29.645789: [debug] Process finished in 4ms: /sbin/ldconfig -p
2020-03-24 14:44:29.646225: [debug] Found shared library libtinfo.so.5 in 'ldconfig -p' output
2020-03-24 14:44:29.647479: [debug] Did not find shared library libtinfo.so.6
2020-03-24 14:44:29.647680: [debug] Did not find shared library libncursesw.so.6
2020-03-24 14:44:29.647842: [debug] Did not find shared library libgmp.so.10
2020-03-24 14:44:29.647951: [debug] Found shared library libgmp.so.3 in 'ldconfig -p' output
2020-03-24 14:44:29.648246: [debug] Potential GHC builds: gmp4
2020-03-24 14:44:29.648378: [debug] Found already installed GHC builds: 
2020-03-24 14:44:29.996438: [debug] Trying to setup GHC build: gmp4
2020-03-24 14:44:29.997594: [error] No setup information found for ghc-8.8.3 on your platform.
This probably means a GHC bindist has not yet been added for OS key 'linux64-gmp4'.
Supported versions: ghc-7.8.4, ghc-7.10.2, ghc-7.10.3, ghc-8.0.1, ghc-8.0.2, ghc-8.2.1, ghc-8.2.2, ghc-8.4.2
```

### Stack version

```
$ stack --version
Version 2.1.3, Git revision 636e3a759d51127df2b62f90772def126cdf6d1f (7735 commits) x86_64 hpack-0.31.2
```

### Method of installation

* Manual install official binary

**Comment by qrilka** [comment]
2 notes:
  * `Not the compiler version we want` says that you have some resolver requiring something else than 8.0.1 - do you run this command in a directory with a `stack.yaml` frile? What `resolver` does it have? If not what resolver do you have in `~/.stack/global-project/stack.yaml`?
 * Other errors seem to come from GHC 7.0.4 in your `/usr/bin` - that one is clearly too old

**Comment by rsolano1-uw** [comment]
Thanks @qrilka, my resolver is `lts-15.5` in the file that you mentioned. I looked at the [stack docs](https://docs.haskellstack.org/en/stable/yaml_configuration/#resolver) and then changed the resolver to `ghc-8.0.1`. This seems to have Stack properly using ghc-8.0.1, but it downloads it instead of using the locally installed version (even if I set `system-ghc: true` in the yaml file).

I'm still having issues with ghc-8.4.4 but it looks like that is caused by my custom-compiled ghc missing some binaries that stack wants. I was able to get ghc-8.4.2 to setup on my CentOS 6 system so I think that this issue is resolved.
